### PR TITLE
SY-4037: Implement Derivative Node in Arc STL

### PR DESCRIPTION
**Pull request by emilbon99**
# Issue Pull Request

## Linear Issue

[SY-4037](https://linear.app/synnaxlabs/issue/SY-4037)

## Description

Adds a `derivative` node to the Arc STL that computes pointwise rate of change (units/second) from input timestamps. Also implements the full C++ `stat` module with `avg`, `min`, `max`, and `derivative` operations, matching the Go-side stat module API.

**Go changes:**
- Added `derivative` to the existing `stat` module alongside `avg`, `min`, `max`
- Derivative symbol uses polymorphic numeric types (matches input type)
- Per-type derivative functions for all 10 numeric types
- 7 new tests (14 total in stat suite)

**C++ changes:**
- New `arc/cpp/stl/stat` module with `Stat` (avg/min/max with reset/duration/count config) and `Derivative` nodes
- Unified `Module` factory handling all four operations
- Registered in the C++ runtime
- 14 tests covering all operations

## Basic Readiness

- [x] I have performed a self-review of my code.
- [x] I have added relevant, automated tests to cover the changes.
- [ ] I have updated documentation to reflect the changes.

<h3>Greptile Summary</h3>

This PR implements a `derivative` node in the Arc STL that computes pointwise rate of change (units/second) across both Go and C++ runtimes, alongside a complete Go `stat` module covering `avg`, `min`, `max`, and `derivative`. It also refactors edge-type resolution into a shared `ResolveNodeTypes` helper, adds UI support in the Console's calculated channel dialog, and extends the `op` code-generation layer with typed derivative functions for all 10 numeric types.

<h3>Confidence Score: 5/5</h3>

Safe to merge; all remaining findings are P2 suggestions that do not affect correctness.

The implementation is complete and well-tested across both runtimes. The derivative output type (always Float64) is correctly propagated from the Go symbol definition through the IR to the C++ state initialisation. The two open findings are a minor C++ test coverage gap for integer-input derivative paths (which are exercised in Go) and a usability regression in graph error messages — neither affects production correctness. Prior review concerns (unsigned UB in C++/Go) were already tracked in previous threads.

arc/cpp/stl/stat/stat_test.cpp — missing integer-input derivative test; arc/go/graph/analyze.go — silent drop of unrecognised edge target params

<details open><summary><h3>Vulnerabilities</h3></summary>

No security concerns identified. The derivative computation operates on in-memory telemetry series with no external input parsing, no secrets handling, and no network exposure. The `xunsafe.CastSlice` usage in the generated op functions is constrained to fixed, known telemetry data types.
</details>

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| arc/cpp/stl/stat/stat.h | New Stat (avg/min/max) and Derivative nodes with full type dispatch; C++ derivative tests only cover the F64 input path |
| arc/go/stl/stat/stat.go | Go stat module with avg/min/max/derivative nodes; derivative always outputs float64 via generated op functions; correct state management across batches |
| x/go/telem/op/gen.go | Adds derivative code-generation template for all 10 numeric types; output is always float64 with state passed in via pointers |
| arc/go/analyzer/resolve.go | New ResolveNodeTypes helper consolidating type unification and substitution; silently skips edges whose target param doesn't exist (intentional per test) |
| arc/go/graph/analyze.go | Replaces local validateEdge + manual unification/substitution steps with shared ResolveNodeTypes; irNodes already carry fresh type variables before the call |
| arc/go/text/analyze.go | Adds ResolveNodeTypes call after building the flow-chain IR; adds Outputs to channel-write node and guard for nodes with no outputs in flow chains |
| core/pkg/service/channel/calculation/analyzer/analyzer.go | Splits DataType into ChanDataType (post-operation, e.g. Float64T for derivative) and ExpressionReturnType; only checks final operation for derivative override |
| core/pkg/service/channel/calculation/compiler/compile.go | PreProcess now runs Analyze to choose function output type based on ExpressionReturnType vs ChanDataType, so the expression function is compiled with the correct pre-operation type |
| arc/cpp/stl/stat/stat_test.cpp | 14 tests covering avg/min/max/derivative; all derivative tests use F64 input only — no coverage for integer input types in the C++ layer |
| arc/go/stl/stat/stat_test.go | Comprehensive Go tests including negative derivatives, cross-batch state, alignment propagation, and integer input types for derivative |

</details>

<h3>Flowchart</h3>

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A[Channel Expression\nexpression return type T] --> B[Arc Expression Function\noutput: ExpressionReturnType]
    B --> C{Operation Type?}
    C -->|avg/min/max| D[Stat Node\noutput: T same type]
    C -->|derivative| E[Derivative Node\noutput: Float64 always]
    D --> F[Channel Write Node\noutput: T]
    E --> G[Channel Write Node\noutput: Float64T]
    F --> H[Synnax Channel\nDataType = T]
    G --> I[Synnax Channel\nDataType = Float64T]
    style E fill:#f9f,stroke:#333
    style G fill:#f9f,stroke:#333
    style I fill:#f9f,stroke:#333
```

<details><summary><h3>Comments Outside Diff (1)</h3></summary>

1. `arc/go/stl/stat/stat.go`, line 89-96 ([link](https://github.com/synnaxlabs/synnax/blob/f13dcb6051c252bdd8fff07e7692a29b22ae59e4/arc/go/stl/stat/stat.go#L89-L96)) 

   <a href="#">P2</a> **Missing `ok` guard for `reductionFn` lookup**

   `reductionMap[inputData.DataType]` does not check whether the data type key exists. If it is absent, `reductionFn` is a nil function and the later call `r.reductionFn(inputData, ...)` in `Next()` will panic. `createDerivative` uses the correct `ok` idiom. Apply the same pattern here:

   ```go
   reductionFn, ok := reductionMap[inputData.DataType]
   if !ok {
       return nil, query.ErrNotFound
   }
   ```
</details>

<sub>Reviews (3): Last reviewed commit: ["Add unit tests for ResolveNodeTypes cove..."](https://github.com/synnaxlabs/synnax/commit/e35823a6fb71e99ff2e8451d09e34fc0b70de8d0) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=27736079)</sub>

### CODE REVIEW

**Comment by emilbon99** [comment]
@greptile please re-review and update PR score

File: console/src/channel/Calculated.tsx
Comment (sy-nico) [inline console/src/channel/Calculated.tsx:1]: Nit: When you select `Derivative` and the window & reset channel control disappear, the Operation group shifts down (it remain in place).

Image

Image

File: arc/go/stl/stat/stat.go [hunk 1]
```diff
@@ -200,41 +231,102 @@ func (r *statNode) Next(ctx node.Context) {
...
+	d.hasPrev = false
+}
+
+func (d *derivativeNode) Next(ctx node.Context) {
+	if !d.RefreshInputs() {
+		return
+	}
+	inputData := d.Input(0)
+	inputTime := d.InputTime(0)
+	if inputData.Len() == 0 {
```
Comment (sy-nico) [inline arc/go/stl/stat/stat.go:319]: Could add a similar check here that `inputTime.Len() == inputData.Len()`. The generated `Derivative<T>` functions index into `inTime[i]` assuming it matches `inputData` length. If `inputTime` is empty or shorter, this panics.

File: arc/cpp/stl/stat/stat.h [hunk 2]
```diff
@@ -0,0 +1,417 @@
...
+
+public:
+    Derivative(runtime::state::Node &&state, types::Kind kind):
+        state(std::move(state)), kind(kind) {}
+
+    x::errors::Error next(runtime::node::Context &ctx) override {
+        if (!this->state.refresh_inputs()) return x::errors::NIL;
+        const auto &input_data = this->state.input(0);
+        const auto &input_time = this->state.input_time(0);
+        if (input_data->size() == 0) return x::errors::NIL;
```
Comment (sy-nico) [inline arc/cpp/stl/stat/stat.h:277]: This checks `input_data->size() == 0` but not `input_time->size()`. `compute<T>` indexes into `input_time->at<int64_t>(i)` assuming it matches `input_data` length.

Add a size check that `input_data->size() == input_time->size()`.

Though this might not be needed given how a series inherently functions.

**Review by sy-nico — CHANGES_REQUESTED** [review]
- [x] Update `functions.mdx` to document the derivative node
- [x] Add integration tests
- [x] Add tests where unsigned integers produce a negative rate
- [x] Add tests for C++ `stat` alignment propagation

File: arc/go/stl/stat/stat.go (hunk 1 above)
Comment (emilbon99) [reply arc/go/stl/stat/stat.go:319]: The `time` series and `data` series are guaranteed to have the same length in arc.

File: arc/cpp/stl/stat/stat.h (hunk 2 above)
Comment (emilbon99) [reply arc/cpp/stl/stat/stat.h:277]: See comment above.

**Review by sy-nico — CHANGES_REQUESTED** [review]
Fix `index out of range [0] with length 0` errors when running on live data. Integration tests (not yet pushed) and manual tests are failing. Already discussed in-person. Simply leaving a note for the PR.

**Comment by emilbon99** [comment]
@greptile there we significant changes. please re-review and update PR score.

**Review by sy-nico — APPROVED** [review]
Approved. Good to merge when tests pass.
